week4-ts/src/components/Header.tsx

```diff
@@ -0,0 +1,178 @@
+    if (!isLocation) {
+      new Promise((resolve) => {
+        navigator.geolocation.getCurrentPosition((currentPosition) => {
+          position.current = currentPosition.coords;
+          const params = {
+            y: position.current.latitude,
+            x: position.current.longitude,
+            radius: 1000,
+            query: '베이커리',
+          };
+          storeSearchHttpHandler(params);
+        });
+      });
+    } else {
+      const params = {
+        y: position.current.latitude,
+        x: position.current.longitude,
+        radius: 1000,
+        query: '베이커리',
+      };
+      storeSearchHttpHandler(params);
+    }
+  };
+
+  const handleInputDisabled = () => {
+    if (null !== searchRef.current) {
+      searchRef.current.disabled = !searchRef.current.disabled;
+      setIsLocation((prev) => !prev);
+    }
+  };
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setInput(e.target.value);
+  };
+  const handleSearchButton = (e: React.MouseEvent<HTMLButtonElement>) => {
+    handleIsSearch(true);
+    e.preventDefault();
+    if (null !== searchRef.current) {
+      if (!searchRef.current.disabled) {
+        const params: Params = {
+          query: input + ' ' + '베이커리',
+        };
+        storeSearchHttpHandler(params);
+      } else {
+        handleMyLocation();
+      }
+    }
+  };
+  return (
+    <StHeaderContainer>
+      <h1>🍰 빵수니가 져아 🍰 </h1>
+      <StHeaderWrapper>
+        <StSearchButton isChoice={isLocation} onClick={handleInputDisabled}>
+          현위치
+        </StSearchButton>
+        <StSearchLabel>우리 동네 </StSearchLabel>
+        <StSearchInput
+          ref={searchRef}
+          type="text"
+          onChange={handleInputChange}
+          value={input}
+          placeholder="지역을 입력해주세요"
+        />
+        <StSearchButton isChoice={isLocation} type="submit" onClick={handleSearchButton}>
+          검색
+        </StSearchButton>
+      </StHeaderWrapper>
+    </StHeaderContainer>
+  );
+}
+export default Header;
+
+interface StHeaderProps {
+  isChoice: boolean;
+}
+const StHeaderContainer = styled.header`
+  ${({ theme }) => {
+    return css`
+      display: flex;
+      flex-direction: column;
+      justify-content: center;
+      width: 100%;
+      height: 100%;
+      padding: 3rem;
+      background-color: ${theme.colors.header};
+
+      color: ${theme.colors.brown};
+    `;
+  }}
+`;
+const StHeaderWrapper = styled.div`
+  ${({ theme }) => {
+    return css`
+      display: flex;
+      justify-content: center;
+      align-items: center;
+      margin: 10px;
+      color: ${theme.colors.brown};
+      font-size: ${theme.fontSizes.form};
+    `;
+  }}
+`;
+const StSearchButton = styled.button`
+  ${({ theme }) => {
+    return css`
+      padding: 30px;
+      margin: 10px;
+      border: none;
+      background: url(${(props: StHeaderProps) => (props.isChoice ? choiceBread : noChoiceBread)});
+      background-repeat: no-repeat;
```

Comment:
SC에서 props 타이핑은 다음과 같이 제네릭으로 넣어줄 수 있어! 

```suggestion
const StSearchButton = styled.button<props: StHeaderProps>`
  ${({ theme }) => {
    return css`
      padding: 30px;
      margin: 10px;
      border: none;
      background: url(${(props) => (props.isChoice ? choiceBread : noChoiceBread)});
      background-repeat: no-repeat;
```

(근데 위 코드처럼 안 해봐서 맞는지 모르겠다)
구조 분해 할당을 사용하면 다음과 같이 되겠지??

```suggestion
const StSearchButton = styled.button<{ isChoice: boolean }>`
  ${({ theme }) => {
    return css`
      padding: 30px;
      margin: 10px;
      border: none;
      background: url(${({ isChice }) => (isChoice ? choiceBread : noChoiceBread)});
      background-repeat: no-repeat;
```

Reply:
```ts
styled.button<StHeaderProps>`
```
이케 해야 데더라구??

구조분해할당 몸에 익히는 중 ㅎㅎㅎ 감쟈감쟈

Reply:
> ```ts
> styled.button<StHeaderProps>`
> ```
> 
> 이케 해야 데더라구??
> 
> 구조분해할당 몸에 익히는 중 ㅎㅎㅎ 감쟈감쟈

이렇게 해야 된다는게 무신 말일까 ?!?